File: tools/source-maltrail/internal/trail/compare.go
package trail

import (
	"crypto/sha256"
	"encoding/hex"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// Diff holds the result of comparing old vs new trail directories.
type Diff struct {
	Changed []string // Relative paths of changed .txt files.
	All     bool     // True when all files should be processed.
}

// Compare walks newDir and diffs each .txt file against oldDir using SHA256.
func Compare(oldDir, newDir string) (*Diff, error) {
	if oldDir == "" || !isDir(oldDir) {
		return &Diff{All: true}, nil
	}

	var changed []string
	err := filepath.WalkDir(newDir, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() || !strings.HasSuffix(strings.ToLower(d.Name()), ".txt") {
			return nil
		}

		rel, err := filepath.Rel(newDir, path)
		if err != nil {
			return err
		}

		oldPath := filepath.Join(oldDir, rel)
		if !isFile(oldPath) || filesDiffer(oldPath, path) {
			changed = append(changed, rel)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}

	sort.Strings(changed)
	return &Diff{Changed: changed}, nil
}

// filesDiffer compares two files by size first, then SHA256 hash.
func filesDiffer(a, b string) bool {
	infoA, errA := os.Stat(a)
	infoB, errB := os.Stat(b)
	if errA != nil || errB != nil {
		return true
	}
	if infoA.Size() != infoB.Size() {
		return true
	}

	hashA, errA := sha256sum(a)
	hashB, errB := sha256sum(b)
	return errA != nil || errB != nil || hashA != hashB
}

// sha256sum returns the hex-encoded SHA256 digest of a file.
func sha256sum(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}
	return hex.EncodeToString(h.Sum(nil)), nil
}

File: v2-connectors/v2-maltrail/src/trail/compare.py
"""SHA256-based comparison of old vs new trail directories."""

from __future__ import annotations

import hashlib
import logging
import os
from dataclasses import dataclass, field
from pathlib import Path

logger = logging.getLogger(__name__)


@dataclass
class Diff:
    """Result of comparing old vs new trail directories."""

    changed: list[str] = field(default_factory=list)  # relative paths of changed .txt files
    all: bool = False  # True when all files should be processed (first run)


def compare(old_dir: str, new_dir: str) -> Diff:
    """Walk new_dir and diff each .txt file against old_dir using SHA256.

    Returns Diff with all=True if old_dir is empty (first run).
    """
    if not old_dir or not os.path.isdir(old_dir):
        return Diff(all=True)

    changed: list[str] = []
    new_path = Path(new_dir)

    for root, _dirs, files in os.walk(new_dir):
        for name in files:
            if not name.lower().endswith(".txt"):
                continue

            full_path = os.path.join(root, name)
            rel = os.path.relpath(full_path, new_dir)
            old_path = os.path.join(old_dir, rel)

            if not os.path.isfile(old_path) or _files_differ(old_path, full_path):
                changed.append(rel)

    changed.sort()
    return Diff(changed=changed)


def _files_differ(a: str, b: str) -> bool:
    """Compare two files by size first, then SHA256 hash."""
    try:
        stat_a = os.stat(a)
        stat_b = os.stat(b)
    except OSError:
        return True

    if stat_a.st_size != stat_b.st_size:
        return True

    return _sha256sum(a) != _sha256sum(b)


def _sha256sum(path: str) -> str:
    """Return hex-encoded SHA256 digest of a file."""
    h = hashlib.sha256()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            h.update(chunk)
    return h.hexdigest()
